File: 算法设计作业/2111454_H5/2111454_H5_Q1.cpp
#include<iostream>

class node
{
public:
    char name;
    node * next;
    node() : name('0') , next(nullptr) {};
    node(char temp) : name(temp) , next(nullptr) {};
};

class nodelist
{
public:
    nodelist();
    void insert(char temp);
    char pop();
    char which(int k,char * cache,int * time);

private:
    int num;
    node * head;
};
nodelist::nodelist()
{
    head = new node();
    num = 0;
}
void nodelist::insert(char temp)
{
    node * p = head;
    node * q = new node(temp);
    for(int i = 0;i < num;i++){
        p = p->next;
    }
    q->next = p->next;
    p->next = q;
    num++;
}
char nodelist::pop()
{
    node * p = head->next;
    head->next = p->next;
    char t = p->name;
    delete p;
    num--;
    return t;
}
char nodelist::which(int k,char * cache,int * time)//找到后续中与当前最远的
{
    node * p = head;
    int * z = new int[k];//存储标志位，以此判断cache中是否有元素不在后续序列中
    for(int i = 0;i < k;i++){//初始化
        z[i] = 0;
    }

    char temp;
    for(int i = 0;i < num;i++){
        p = p->next;
        for(int j = 0;j < k;j++){
            if(p->name == cache[j] && z[j] == 0){//当序列中的数和cache一样时，将temp赋值为序列中的字符，并标志位赋值为1
                temp = p->name;
                z[j] = 1;
            }
        }
    }

    int flag = 0;
    for(int i = 0;i < k;i++){//探查是否存在不再后续序列中的cache字符
        if(z[i] == 0 && flag == 0){
            flag = i+1;
        }
        else if(z[i] == 0 && time[i] > time[flag-1]){//当出现多个时，比较时间，将时间较大的进行赋值
            flag = i+1;
        }
    }

    if(flag != 0){
        return cache[flag-1];
    }
    else{
        return temp;
    }
}

void addtime(int k,int * time){//增加时间
    for(int i = 0;i < k;i++){
        time[i]++;
    }
}

int main()
{
    int k;//共有多少cache
    int n;
    int s;//序列共有多少
    int no;
    nodelist q;
    std::cin>>k>>n>>s;
    char * cache = new char[k];
    int * time = new int[k];//存储未被使用的时间
    char ans[1000];
    int ans_num = 0;
    std::cin>>no;
    for(int i = 0;i < s;i++){
        char temp;
        std::cin>>temp;
        q.insert(temp);
    }

    for(int i = 0;i < k;i++){//初始化time、cache
        time[i] = 0;
        cache[i] = 0;
    }
    for(int i = 0;i < k;i++){//最初时直接将序列中的字符存入cache
        cache[i] = q.pop();
        for(int i = 0;i < k;i++){
            if(cache[i] != 0){
                time[i]++;
            }
        }
        s--;
    }
    while(s != 0){
        char temp = q.which(k,cache,time);
        char temp2 = q.pop();
        bool flag = true;
        for(int i = 0;i < k;i++){//当序列中的字符在cache中存在时，跳出
            if(temp2 == cache[i]){
                time[i] = 0;
                flag = false;
                break;
            }
        }
        if(flag){
            for(int i = 0;i < k;i++){//当序列中的字符等于cache中的字符，将二者替换，同时重置时间
                if(temp == cache[i]){
                    ans[ans_num] = cache[i];
                    ans_num++;
                    cache[i] = temp2;
                    time[i] = 0;
                    break;
                }
            }
        }
        addtime(k,time);
        s--;
    }

    for(int i = 0;i < ans_num;i++){
        if(i == 0){
            std::cout<<ans[i];
        }
        else{
            std::cout<<" "<<ans[i];
        }
    }
}

File: 算法设计作业/2111454_H5/python版本/2111454_H5_Q1.py
def which(k,cache,time,pls):
    ta = -1
    a = -1
    b = -1
    c = [0 for x in range(k)]

    for i in range(k):
        if cache[i] not in pls and time[i] > ta: #当cache中的值不在pls中且对应未使用时间比原先的大时，将此时cache的下标赋值给a
            a = i
            ta = time[i]
        elif cache[i] in pls and c[i] == 0 and b < pls.index(cache[i]): 
            """当cache中的值在pls中且此时cache的值未被记录,
               同时b小于当前cache元素在pls中对应的下标时,
               将pls的下标赋值给b"""
            c[i] = 1
            b = pls.index(cache[i])
    
    if a != -1:
        return a , False
    else:
        return b , True

def timeadd(k,time):
    for i in range(k):
        time[i] += 1
    return time

if __name__ == '__main__':
    k , n , s = map(int, input().split()) #cache的总个数
    no = input()
    cache = []
    time = [0 for x in range(k)] #存储多长时间未使用对应cache
    ans = [] #存储答案
    
    pls1 = input()
    pls2 = pls1.replace(' ', '')
    pls = list(pls2)
    
    t = len(pls)
    while t != 0:
        if s - t < k: #将请求序列pls的前k个元素依次赋值给k个cache
            add = pls[0]
            cache.append(add)
            pls.remove(add)
            t -= 1
        else:
            target,flag = which(k,cache,time,pls)
            if pls[0] in cache: #当pls[0]在cache中时
                temp = pls[0]
                time[cache.index(temp)] = 0
                t -= 1
                pls.remove(temp)
                time = timeadd(k,time)
                continue
            if flag: #cache中的元素都可以在pls中找到,且pls不在cache时
                temp = pls[target] #将请求序列pls中的目标元素值赋值给temp
                temp2 = pls[0] #将将要插入的请求序列pls的值赋值给temp2
                re_index = cache.index(temp) #记录下需要替换的cache的下标
                time[re_index] = 0 #重置time计数器的值
                ans.append(cache[re_index])
                cache[re_index] = temp2 #替换cache的值
                pls.remove(temp2)
                t -= 1
                time = timeadd(k,time)
            else: #cache中的元素出现了不能在pls中找到
                temp = pls[0] #将将要插入的请求序列pls的值赋值给temp
                time[target] = 0 #重置time计时器的值
                ans.append(cache[target]) 
                cache[target] = temp#替换cache的值
                pls.remove(temp)
                t -= 1
                time = timeadd(k,time)

    for i in range(0,len(ans)):
        print(ans[i], end=" ")